Revised "asyc_scope" section

lucteo opened the pull request:
- rename section to "Async Scope, usage guide"
- made the subsections provide more of an overview; move technical details to specification section
- provide usage examples for all operations
- put `nest` after `spawn` and `spawn_future`
- some fixes in the specification section

lucteo committed:
Revised "asyc_scope" section

- rename section to "Async Scope, usage guide"
- made the subsections provide more of an overview; move technical details to specification section
- provide usage examples for all operations
- put `nest` after `spawn` and `spawn_future`
- some fixes in the specification section

kirkshoop commented on asyncscope.md:
```diff
@@ -148,84 +150,141 @@ struct async_scope {
 
-Connecting and starting the `nested_sender<Values...>`, `n`, returned from `nest()` will connect and start `s`.
+The given sender must complete with `void` or be stopped.
+It is not accepted a sender that can complete with errors; the user must explicitly handle the errors that might appear before passing the corresponding sender to `spwawn`.
```
spwawn -> spawn

kirkshoop commented on asyncscope.md:
````diff
@@ -148,84 +150,141 @@ struct async_scope {
+async_scope s;
+for (int i=0; i<100; i++)
+    s.spawn(on(sched, some_work(i)));
+```
````
I am wondering if this should do something with `on_empty()` to show that s must stay in scope until the work is complete

kirkshoop commented on asyncscope.md:
```diff
@@ -148,84 +150,141 @@ struct async_scope {
-`s` is guaranteed to `start()` if allocation of the `operation_state` succeeds.
-`s` is not required to `start()` before `spawn()` returns.
+Eagerly launches work on the `async_scope` but returns a `@_spawn-future-sender_@` that represents an eagerly running task.
+This involves an allocation for the `operation_state` of the sender, until it completes, and synchronization to resolve the race between the production of the result and the consumption of the result.
```
I think that the feedback would include changing `operation_state` to `@_operation-state_@`

kirkshoop commented on asyncscope.md:
```diff
@@ -148,84 +150,141 @@ struct async_scope {
+`template <sender S> @_nest-sender_@<S> nest(S&& s);`
 
-`spawn_future(sender) const&->future_sender<Values…>;`
+Returns a `nested_sender` that extends the lifetime of the `async_scope` that produced it to include the lifetime of the nested_sender object and, if started, the lifetime of the given sender operation.
```
I think I saw more than one place where `nested_sender` was not changed to _`nest-sender`_

kirkshoop commented on asyncscope.md:
```diff
@@ -148,84 +150,141 @@ struct async_scope {
-To safely destroy the `async_scope` object an empty_sender start must happen-after all calls to `nest()`, `spawn()` and `spawn_future()` are complete. Once empty_sender completes the async_scope can be safely destructed - or a new set of calls to `nest()`, `spawn()` and `spawn_future()` happen-before the start of a new empty_sender and the async_scope destructor happens-after the completion of th ...
+To safely destroy the `async_scope` object it's recommended to use `@_on-empty-sender_@` to get notified when the scope object finished executing all the work.
+In this case, the `@_on-empty-sender_@` start must happen-after all calls to `nest()`, `spawn()` and `spawn_future()` are complete.
+Moreover, no more calls to `nest()`, `spawn()` and `spawn_future()` should be made after the call to `on_empty()`.
```
`on_empty()` does not participate in a race. It is `start()` of the sender `on_empty()` returns that participates in the race.

kirkshoop commented on asyncscope.md:
```diff
@@ -234,22 +293,55 @@ That is to say the following is safe:
+  for ( int i=0; i<num_jobs; i++ )
+    scope.spawn(on(sched, some_work(i)));
+  // Join the work with the help of the scope
+  return scope.empty();
```
`empty()` -> `on_empty()`

kirkshoop commented on asyncscope.md:
````diff
@@ -234,22 +293,55 @@ That is to say the following is safe:
+    // all work, static or dynamic is completed at this point
+    return 0;
+}
+```
````
I do not see any use of `on_empty()`, is that intentional?

lucteo replied:
no, it was a mistake on my part

kirkshoop commented on asyncscope.md:
```diff
@@ -136,7 +136,9 @@ The requirements for the async scope are:
  - An `async_scope` must constrain `spawn()` and `spawn_future()` to accept only senders that are never-blocking.
  - An `async_scope` must constrain `spawn()` to accept only senders that complete with void.
- - An `async_scope` must provide an `empty_sender` that completes when all spawned senders are complete.
+ - An `async_scope` must provide an `@_on-empty-sender_@` that completes when all spawned senders are complete.
```
One of the pieces of feedback was to add a bullet for the requirement that `spawn()` and `spawn_future()` must start the given sender before they exit.

lucteo committed:
Incorporated review feedback

lucteo committed:
Merge remote-tracking branch 'origin/main' into async_scope_section

lucteo commented on asyncscope.md:
```diff
@@ -318,84 +321,145 @@ struct async_scope {
 ## Lifetime
 
-An `async_scope` object must outlive work that is spawned on it. It should be viewed as owning the storage for that work.
+An `async_scope` object must outlive work that is spawned on it.
```
Note: everywhere I can, I split paragraphs into multiple lines, so that we have one sentence per line. This makes the diffing much easier with a lot of tools.

Please excuse me if there are too many changes for things that shouldn't be

lucteo commented on asyncscope.md:
```diff
@@ -318,84 +321,145 @@ struct async_scope {
 
-An `async_scope` object must outlive work that is spawned on it. It should be viewed as owning the storage for that work.
+An `async_scope` object must outlive work that is spawned on it.
+It should be viewed as owning the storage for that work.
```
Note: This phrase made my mind think it's better to drop constness from `spawn`, `spawn_future` and `nest`.

lucteo committed:
Minor fixes

- clarification on what *empty* means
- spec changes to say that operations are started when spawning
- spec changes for `get_stop_token()`
- other minor fixes

lucteo committed:
Work on never-blocking

- remove the requirement line, as we cannot achieve it
- strengthen the description in the subsection dedicated to `spawn()`

lucteo committed:
Formatting changes in Q & A section

lucteo committed:
Minor changes to the design decisions part

lucteo committed:
Tidy up the code snippets

- section 4 definitions are now in code blocks
- copy `async_scope` definition from specification to section 4, but without the additional classes.
- arrange the `serializer` code
- rename `async-scope-receiver` into `spawn-receiver`
